File: lab-redes/python/websocket/mural_servidor.py
import asyncio
import os
import sys

import websockets

PORTA_BASE = 8888
clientes_conectados = set()


def ler_offset() -> int:
    if len(sys.argv) > 1 and sys.argv[1].isdigit():
        return int(sys.argv[1])
    return int(os.getenv("LAB_REDES_OFFSET", "0"))


async def tratar_conexao(websocket):
    clientes_conectados.add(websocket)
    print(f"[WebSocket] Novo aluno conectado. Total: {len(clientes_conectados)}", flush=True)
    await websocket.send("Bem-vindo(a) ao mural de avisos da turma!")

    try:
        async for mensagem in websocket:
            print(f"[WebSocket] Recebido: {mensagem}", flush=True)
            aviso_formatado = f"Aviso da turma: {mensagem}"
            websockets.broadcast(clientes_conectados, aviso_formatado)
    finally:
        clientes_conectados.discard(websocket)
        print(f"[WebSocket] Aluno desconectado. Total: {len(clientes_conectados)}", flush=True)


async def main():
    porta = PORTA_BASE + ler_offset()
    print(f"[WebSocket] Servidor do mural iniciado na porta {porta}.", flush=True)
    async with websockets.serve(tratar_conexao, "0.0.0.0", porta):
        await asyncio.Future()


asyncio.run(main())

File: lab-redes/java/websocket/src/main/java/MuralServidor.java
import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;
import org.java_websocket.server.WebSocketServer;

import java.net.InetSocketAddress;

public class MuralServidor extends WebSocketServer {
    private static final int PORTA_BASE = 8887;

    public MuralServidor(int porta) {
        super(new InetSocketAddress(porta));
    }

    @Override
    public void onOpen(WebSocket conexao, ClientHandshake handshake) {
        System.out.println("[WebSocket] Novo aluno conectado: " + conexao.getRemoteSocketAddress());
        conexao.send("Bem-vindo(a) ao mural de avisos da turma!");
    }

    @Override
    public void onMessage(WebSocket conexao, String mensagem) {
        System.out.println("[WebSocket] Recebido: " + mensagem);
        String avisoFormatado = "Aviso da turma: " + mensagem;

        for (WebSocket cliente : getConnections()) {
            cliente.send(avisoFormatado);
        }
    }

    @Override
    public void onClose(WebSocket conexao, int codigo, String motivo, boolean remoto) {
        System.out.println("[WebSocket] Aluno desconectado: " + conexao.getRemoteSocketAddress());
    }

    @Override
    public void onError(WebSocket conexao, Exception ex) {
        System.out.println("[WebSocket] Erro: " + ex.getMessage());
    }

    @Override
    public void onStart() {
        System.out.println("[WebSocket] Servidor do mural iniciado.");
    }

    public static void main(String[] args) {
        int porta = PORTA_BASE + lerOffset(args);
        MuralServidor servidor = new MuralServidor(porta);
        servidor.start();
        System.out.println("[WebSocket] Escutando na porta " + porta + ".");
    }

    private static int lerOffset(String[] args) {
        if (args.length > 0 && args[0].matches("\\d+")) {
            return Integer.parseInt(args[0]);
        }

        String valor = System.getenv("LAB_REDES_OFFSET");
        return valor == null || valor.isBlank() ? 0 : Integer.parseInt(valor);
    }
}
